---
title: "Phylogenetic Tree Inspection & Rooting"
author: "Maya Craig"
date: '`r Sys.Date()`'
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/03_Phylogenetic_Tree/",
                      warning = FALSE) 
```


# Goals 

1. Load the fastree unrooted tree.  
2. Add tree to phyloseq object.  
3. Visualize and inspect tree with ggtree. 
4. Prune ASVs, if needed.  
5. Root our tree. 
6. Combine new tree with a phyloseq object. 
7. Save 2 phyloseq objects: 1. Unrooted tree phyloseq object, 2. Rooted tree phyloseq object.


# Before you start

## Set my seed 
```{r set-seed}
set.seed(98572)
```

## Load Packages 
```{r load-packages}
pacman::p_load(tidyverse, phyloseq, ggtree, phytools,
               install = FALSE)
```

## Load Data files 
```{r load-data}
# Preprocessed phyloseq object 
load("data/02_PreProcessing/raw_physeq.RData")
raw_physeq

# Load in the tree! 
unrooted_tree <- read.tree("data/03_Phylogenetic_Tree/ASVs_unrooted.tree")
unrooted_tree
str(unrooted_tree)
```


# Merge Phyloseq 
```{r merge-physeq}
# Intuition check 
stopifnot(ntaxa(raw_physeq) == ntaxa(unrooted_tree))

# Merge the tree with the phyloseq object 
unrooted_physeq <- 
  merge_phyloseq(raw_physeq, unrooted_tree)
unrooted_physeq
```

# Plot Tree with `ggtree`

```{r plot-tree-unrooted}
# Make a basic tree
kingdom_tree <- 
  ggtree(unrooted_physeq) + 
  # color tips by kingdom 
  geom_tippoint(mapping = aes(color = Kingdom)) +
  scale_color_manual(values = c("goldenrod1", "cornflowerblue", "grey")) +
  # Add title 
  labs(title = "Unrooted Tree") + 
  #move the legend to the bottom 
  theme(legend.position = "bottom"); kingdom_tree

kingdom_node_tree <- 
  kingdom_tree + 
  # Add the node label 
  geom_text(aes(label=node), hjust= -0.5, vjust = -0.3, size = 1)
kingdom_node_tree
```
- It looks like there are two very long branches that need to be further investigated. There are also quite a few NAs.

# Evaluate Long Branch 

```{r eval-long-branch}
# View a specific clade 
viewClade(kingdom_node_tree +
          labs(title = "Unrooted Tree: Node 2955"), 
          node = 5181)

viewClade(kingdom_node_tree + 
          labs(title = "Unrooted Tree: Node 2955") + 
          geom_text(aes(label=ASV)), 
          node = 2995)
#Other clades that were investigating in attempt to isolate the suspicious outlier: 5181, 2995, 2846, 5846, 3658, 7854
```
- ASV 4113, 4452, and 4425 are outliers on the tree (Mar and I used iTol to further investigate the outliers because it was difficult to subset by clade here in RStudio)

```{r check-taxonomy}
unrooted_physeq %>%
  subset_taxa(., ASV == "ASV_4113" | ASV == "ASV_4425" | ASV == "ASV_4452") %>%
  tax_table() %>%
  data.frame()

# Let's also check the counts of the ASV 
unrooted_physeq %>%
  subset_taxa(., ASV == "ASV_4113" | ASV == "ASV_4425" | ASV == "ASV_4452") %>%
  otu_table() %>%
  data.frame() %>%
  colSums()
```

ASV 4113 belongs to the Thermoactinomycetaceae Family and the Kroppenstedtia Genus. Two of this ASV is found in one sample (sample SRR23008990) but not found in any of the other samples. I will BLAST the sequence to find out more information!

After a BLAST search, I found that ASV 4113 matched an uncultured clone of Kloppenstedtia 16 rRNA gene with an E value of 4e-132 and a percent identify of 99.62%. This appears to be a contaminant, so this will be removed 

The BLAST search showed that ASV_4425 belong to a carp (fish?) genome, this is probably a contaminate and will be removed. 

The BLAST search showed that ASV_4452 100% identifty to a strain of coronavirus. This will be removed as well.

# Prune ASV_4425, ASV_4452, ASV_4113
```{r prune-ASV456}
# Function from Joey McMurdie: https://github.com/joey711/phyloseq/issues/652
pop_taxa = function(physeq, badTaxa){
  allTaxa <-  taxa_names(physeq)
  allTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
  return(prune_taxa(allTaxa, physeq))}

# Let's use the pop_taxa function :) 
# Recreate a phyloseq object without ASV_4452, 4425, 4113
unrooted_physeq_rmASVs <- 
  unrooted_physeq %>%
  pop_taxa(., c("ASV_4425","ASV_4452", "ASV_4113" ))

# Intuition Check
ntaxa(unrooted_physeq) - ntaxa(unrooted_physeq_rmASVs) 

# Visually inspect 
ggtree(unrooted_physeq_rmASVs) + 
  geom_tippoint(mapping = aes(color = Kingdom))
```
- After inspection, the outlier ASVs were successfully removed from the tree.

# Midroot Tree
```{r midroot-tree}
# Is the tree rooted?
new_unrooted_tree <-phy_tree(unrooted_physeq_rmASVs)
is.rooted(new_unrooted_tree)

# Let's midpoint root the tree
midpoint_rooted_tree <- midpoint.root(new_unrooted_tree)

# Is the new tree rooted?
is.rooted(midpoint_rooted_tree)

# Assign to a new phyloseq object: merging subsetted phyloseq with the new rooted tree

# 1. Create the phyloseq object without contaminated ASVs
physeq_rmASVs <- 
  raw_physeq %>%
  subset_taxa(ASV != c("ASV_4425","ASV_4452","ASV_4113"))

# Merge tree with the new physeq_rm456
midroot_physeq_rmASVs <- 
  merge_phyloseq(physeq_rmASVs, midpoint_rooted_tree)
midroot_physeq_rmASVs

# Quick inspection of tree 
ggtree(midroot_physeq_rmASVs) + 
  geom_tippoint(mapping = aes(color = Kingdom))
```

# Save to a new phyloseq object
```{r save-physeq}
# Save phyloseq objects with our tree object to one .RData file 
save(list = c("unrooted_physeq_rmASVs", "midroot_physeq_rmASVs"),
     file = "data/03_Phylogenetic_Tree/phytree_preprocessed_physeq.RData")
```

# Document conclusions 
- The Bai et al paper did not asses phylogenetics and did not describe how contaminate ASVs were removed. In my analysis, I was able to successfully produce a phylogenetic tree using data that was aligned by MAFT and assembled by FastTree2.

# Session Information 
```{r session-info}
devtools::session_info()
```
